### ISSUE TITLE
ipynb reader performance: JSON escaped double quote quadratic in time

### ISSUE THREAD

**Issue by ickc** [jgm/pandoc#7744]
TL;DR: skip to the end that this is probably a O(n^1.778) time in the number of `\"` in a JSON string.

# Minimal working example

In the following notebook, it runs a single cell

```python
import holoviews as hv
hv.extension("matplotlib")
```

[holoview-load-only.ipynb.zip](https://github.com/jgm/pandoc/files/7689482/holoview-load-only.ipynb.zip)

Using the above notebook, run the following:

```bash
pandoc -s -o holoview-load-only-pandoc.html holoview-load-only.ipynb
```

And the process takes forever to finish.

However, running it through nbconvert,

```bash
jupyter-nbconvert holoview-load-only.ipynb --to html
```

it is fine.

# An even more minimal working example

In this file, we take the above file and manually removed some data, until it has a single very long element that confuses pandoc:

[holoview-load-only-manual_filtered.ipynb.zip](https://github.com/jgm/pandoc/files/7689484/holoview-load-only-manual_filtered.ipynb.zip)

Note the data with key `"application/vnd.holoviews_load.v0+json"`. This does not confuses nbconvert (which probably uses Python's JSON parser) but, e.g., also confuses textmate.

Observation: probably the excessive amount of escaped double quote `\"` is the culprit.

# An actually minimal working example

Replace the long string by a bunch of escaped double quote:

[holoview-load-only-manual_4s.ipynb.zip](https://github.com/jgm/pandoc/files/7689487/holoview-load-only-manual_4s.ipynb.zip)

i.e. our really minimal working example is

```json
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.holoviews_load.v0+json": "\"\"\"...add_more_\"_here..."
  	},
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import holoviews as hv\n",
    "hv.extension(\"matplotlib\")"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "text_representation": {
    "extension": ".py",
    "format_name": "percent",
    "format_version": "1.3",
    "jupytext_version": "1.13.3"
   }
  },
  "kernelspec": {
   "display_name": "all39-defaults",
   "language": "python",
   "name": "all39-defaults"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
```

# Benchmarking the performance

Using this idea and study how time depends on the length of `\"` in a string,
we found that it is quadratic time:

([notebook used here](https://github.com/ickc/pandoc-ipynb/blob/master/pandoc-performance.ipynb))

```python
from copy import deepcopy
import timeit
import json
from pathlib import Path
from functools import partial

from panflute import convert_text
import numpy as np
import plotly.express as px
import pandas as pd
import statsmodels.api as sm
from map_parallel import map_parallel
```

```python
# pandoc commit 6b962e6b0e729e925c4313427926cc6bfc5d3d1f
pandoc_path = Path("~/.local/bin/pandoc").expanduser()
# input ipynb
path = Path("holoview-load-only-manual_filtered.ipynb").expanduser()
```

```python
!$pandoc_path --version
```

    pandoc 2.16.2
    Compiled with pandoc-types 1.22.1, texmath 0.12.3.3, skylighting 0.12.1,
    citeproc 0.6, ipynb 0.2
    User data directory: ~/.local/share/pandoc
    Copyright (C) 2006-2021 John MacFarlane. Web:  https://pandoc.org
    This is free software; see the source for copying conditions. There is no
    warranty, not even for merchantability or fitness for a particular purpose.

```python
with path.open("r") as f:
    data = json.load(f)
```

```python
n_original = len(data["cells"][0]["outputs"][0]["data"]["application/vnd.holoviews_load.v0+json"])
n_original
```

    1679380

```python
data["cells"][0]["outputs"][0]["data"]["application/vnd.holoviews_load.v0+json"] = '"'
```

```python
def cook_new_data(data, n):
    res = deepcopy(data)
    res["cells"][0]["outputs"][0]["data"]["application/vnd.holoviews_load.v0+json"] = '"' * n
    return res
```

```python
def timethis(
    data,
    n,
    number=1,
    repeat=1,
):
    return np.mean(
        timeit.repeat(
            'convert_text(text, input_format="ipynb", output_format="native", pandoc_path=pandoc_path)',
            setup='import json; from panflute import convert_text; text = json.dumps(cook_new_data(data, n))',
            globals={"data": data, "n": n, "cook_new_data": cook_new_data, "pandoc_path": pandoc_path},
            number=number,
            repeat=repeat,
        )
    )
```

```python
N = 18
n = 2**np.arange(1, N)
```

```python
res = map_parallel(partial(timethis, data), n, mode="multithreading", processes=len(n))
```

```python
df = pd.DataFrame(
    {
        "n": n,
        "time (s)": res,
    }
)
```

```python
df
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>n</th>
      <th>time (s)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2</td>
      <td>0.073833</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4</td>
      <td>0.099905</td>
    </tr>
    <tr>
      <th>2</th>
      <td>8</td>
      <td>0.051142</td>
    </tr>
    <tr>
      <th>3</th>
      <td>16</td>
      <td>0.057677</td>
    </tr>
    <tr>
      <th>4</th>
      <td>32</td>
      <td>0.087921</td>
    </tr>
    <tr>
      <th>5</th>
      <td>64</td>
      <td>0.055302</td>
    </tr>
    <tr>
      <th>6</th>
      <td>128</td>
      <td>0.107645</td>
    </tr>
    <tr>
      <th>7</th>
      <td>256</td>
      <td>0.055969</td>
    </tr>
    <tr>
      <th>8</th>
      <td>512</td>
      <td>0.098900</td>
    </tr>
    <tr>
      <th>9</th>
      <td>1024</td>
      <td>0.102706</td>
    </tr>
    <tr>
      <th>10</th>
      <td>2048</td>
      <td>0.149745</td>
    </tr>
    <tr>
      <th>11</th>
      <td>4096</td>
      <td>0.312231</td>
    </tr>
    <tr>
      <th>12</th>
      <td>8192</td>
      <td>0.989595</td>
    </tr>
    <tr>
      <th>13</th>
      <td>16384</td>
      <td>3.685738</td>
    </tr>
    <tr>
      <th>14</th>
      <td>32768</td>
      <td>14.235585</td>
    </tr>
    <tr>
      <th>15</th>
      <td>65536</td>
      <td>53.312253</td>
    </tr>
    <tr>
      <th>16</th>
      <td>131072</td>
      <td>197.977269</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.plot(backend="plotly", x="n", y="time (s)", log_x=True, log_y=True)
```

![newplot](https://user-images.githubusercontent.com/10970208/145505873-a22a7456-c03e-41db-96fd-cd9df8c257b0.png)

```python
df_fit = df.iloc[10:]
```

$$t = A n^x$$

$$\log t = C + x \log n$$

```python
X = sm.add_constant(np.log(df_fit["n"]))
y = np.log(df_fit["time (s)"])
model = sm.OLS(y, X)
results = model.fit()
results.summary()
```

    ~/.mambaforge/envs/all39-defaults/lib/python3.9/site-packages/statsmodels/stats/stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 7 samples were given.
      warn("omni_normtest is not valid with less than 8 observations; %i "

<table class="simpletable">
<caption>OLS Regression Results</caption>
<tr>
  <th>Dep. Variable:</th>        <td>time (s)</td>     <th>  R-squared:         </th> <td>   0.994</td>
</tr>
<tr>
  <th>Model:</th>                   <td>OLS</td>       <th>  Adj. R-squared:    </th> <td>   0.993</td>
</tr>
<tr>
  <th>Method:</th>             <td>Least Squares</td>  <th>  F-statistic:       </th> <td>   881.7</td>
</tr>
<tr>
  <th>Date:</th>             <td>Thu, 09 Dec 2021</td> <th>  Prob (F-statistic):</th> <td>8.12e-07</td>
</tr>
<tr>
  <th>Time:</th>                 <td>18:03:06</td>     <th>  Log-Likelihood:    </th> <td>  1.8560</td>
</tr>
<tr>
  <th>No. Observations:</th>      <td>     7</td>      <th>  AIC:               </th> <td>  0.2879</td>
</tr>
<tr>
  <th>Df Residuals:</th>          <td>     5</td>      <th>  BIC:               </th> <td>  0.1797</td>
</tr>
<tr>
  <th>Df Model:</th>              <td>     1</td>      <th>                     </th>     <td> </td>   
</tr>
<tr>
  <th>Covariance Type:</th>      <td>nonrobust</td>    <th>                     </th>     <td> </td>   
</tr>
</table>
<table class="simpletable">
<tr>
    <td></td>       <th>coef</th>     <th>std err</th>      <th>t</th>      <th>P>|t|</th>  <th>[0.025</th>    <th>0.975]</th>  
</tr>
<tr>
  <th>const</th> <td>  -15.8036</td> <td>    0.587</td> <td>  -26.924</td> <td> 0.000</td> <td>  -17.312</td> <td>  -14.295</td>
</tr>
<tr>
  <th>n</th>     <td>    1.7780</td> <td>    0.060</td> <td>   29.693</td> <td> 0.000</td> <td>    1.624</td> <td>    1.932</td>
</tr>
</table>
<table class="simpletable">
<tr>
  <th>Omnibus:</th>       <td>   nan</td> <th>  Durbin-Watson:     </th> <td>   1.198</td>
</tr>
<tr>
  <th>Prob(Omnibus):</th> <td>   nan</td> <th>  Jarque-Bera (JB):  </th> <td>   0.582</td>
</tr>
<tr>
  <th>Skew:</th>          <td> 0.601</td> <th>  Prob(JB):          </th> <td>   0.748</td>
</tr>
<tr>
  <th>Kurtosis:</th>      <td> 2.260</td> <th>  Cond. No.          </th> <td>    70.0</td>
</tr>
</table><br/><br/>Notes:<br/>[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.

Predicted time

```python
np.exp(results.params.const + results.params.n * np.log(n_original))
```

    16027.401232530488

# Potentially related issues

- #5166
- #6964
- #7306

**Comment by jgm** [comment]
The problem doesn't seem to be one of parsing the structure into an ipynb notebook (what the `ipynb` package does).  The following test program finishes almost immediately:
```hs
{-# LANGUAGE ScopedTypeVariables #-}
import Data.Ipynb
import Data.ByteString.Lazy as B
import Data.Aeson

main = do
  h <- B.readFile "./holoview.ipynb"
  let Just (nb :: Notebook NbV4) = decode h
  print nb
```
(where holoview.ipynb is from your second case above).

**Comment by jgm** [comment]
It seems to have to do with the `v0+json` part of the mime type; when I replace that with `v0` it finishes fast.

**Comment by jgm** [comment]
What does the `+json` mean here? Looks like the contents are javascript code.

**Comment by jgm** [comment]
`ipynb` is marking it a JsonData, which seems to be what it's supposed to do when the mime type has a `+json` suffix.  So the `ipynb` reader returns this as an aeson Value, which is just a String.

Pandoc's reader sees this and will `encode` the Value back to a JSON string, which it inserts into the mediabag.

So far, though, I don't see any trouble spots.

**Comment by jgm** [comment]
I don't see how those other issues can be relevant, since in this case all the content is in something that just becomes a code block (and so isn't parsed as markdown).

**Comment by ickc** [comment]
I looked a bit into the `+json`:

```json
    "application/vnd.exampleorg.type+json": {
      # JSON data, included as-is, when the mime-type key ends in +json
```

from <https://nbformat.readthedocs.io/en/latest/format_description.html#display-data>.

I tried using another example with `+json`, not the holoview example above, but with plotly directly ([just-plotly.ipynb.zip](https://github.com/jgm/pandoc/files/7691334/just-plotly.ipynb.zip) from #7745.) I find that manually removing the `"application/vnd.plotly.v1+json"` key has no effect as far as using nbconvert to convert from ipynb to html is concerned.

These data is very likely used by the respective `exampleorg` as private data in a live jupyter(lab) section to pass data around.

So the idea would be to treat these `+json` data specially. If not `--ipynb-output=all`, strip them. Else, I guess until we know where the slow down comes from and can fix them, may be still strip them in the interim.

What do you think?

**Comment by ickc** [comment]
@jgm, I look more into `application/json` in addition to `"^application/(.*\\+)?json$"`. In short, nbconvert completely ignore them when converting to html. Currently, pandoc will put `application/json` into a CodBlock with class json. The proposal is to either completely strip them in ipynb reader, or at least drop them in writers other than ipynb writer.

More details:

Inside [just-mock-json.zip](https://github.com/jgm/pandoc/files/7696386/just-mock-json.zip), you'll find a ipynb file with 2 html outputs produced by

```bash
jupyter-nbconvert just-mock-json.ipynb --to html
pandoc -s -o just-mock-json-pandoc.html just-mock-json.ipynb -V header-includes='<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js"></script>' --extract-media=media
```

The one produced by nbconvert has the `application/json` stripped out. While pandoc display it as a codeblock.

**Comment by jgm** [comment]
If we completely strip them in the ipynb reader, then we lose round-trip capacity.
We could strip them in `filterIpynbOutput`, unless `--ipnb-output=all` is specified -- maybe that's what you had in mind.

**Comment by ickc** [comment]
> We could strip them in `filterIpynbOutput`, unless `--ipnb-output=all` is specified -- maybe that's what you had in mind.

I mentioned that as a possibility too, but also caveated that at least for now we know it can takes O(n^2)-ish to finish for some cases.
